hydroflow_lang/src/graph/ops/group_by.rs

```diff
@@ -31,32 +67,102 @@ pub const GROUP_BY: OperatorConstraints = OperatorConstraints {
     ports_out: None,
     num_args: 2,
     input_delaytype_fn: &|_| Some(DelayType::Stratum),
-    write_fn: &(|&WriteContextArgs { op_span, .. },
-                 &WriteIteratorArgs {
+    write_fn: &(|wc @ &WriteContextArgs { op_span, .. },
+                 wi @ &WriteIteratorArgs {
                      ident,
                      inputs,
+                     generic_args,
                      arguments,
                      is_pull,
+                     op_name,
                      ..
                  },
-                 _| {
+                 diagnostics| {
         assert!(is_pull);
+
+        let generics_span = generic_args.map(Spanned::span).unwrap_or(op_span);
+
+        let persistence = parse_persistence_lifetimes(wi, diagnostics);
+        let persistence = match *persistence {
+            [] => Persistence::Static,
+            [a] => a,
+            _ => {
+                diagnostics.push(Diagnostic::spanned(
+                    generics_span,
+                    Level::Error,
+                    format!(
+                        "Operator `{}` expects zero or one persistence lifetime generic arguments",
+                        op_name
+                    ),
+                ));
+                Persistence::Static
+            }
+        };
+
+        let mut generic_type_args = parse_generic_types(wi);
+        if !generic_type_args.is_empty() && 2 != generic_type_args.len() {
+            diagnostics.push(Diagnostic::spanned(
+                generics_span,
+                Level::Error,
+                format!("Operator `{}` expects zero or two type arguments", op_name),
+            ));
+            generic_type_args.clear();
+        }
+
         let input = &inputs[0];
         let initfn = &arguments[0];
         let aggfn = &arguments[1];
-        let write_iterator = quote_spanned! {op_span=>
-            let #ident = {
-                #[inline(always)]
-                fn check_input<Iter: ::std::iter::Iterator<Item = (A, B)>, A, B>(iter: Iter) -> impl ::std::iter::Iterator<Item = (A, B)> { iter }
-                check_input(#input).fold(::std::collections::HashMap::new(), |mut ht, nxt| {
-                    let e = ht.entry(nxt.0).or_insert_with(#initfn);
-                    #[allow(clippy::redundant_closure_call)]
-                    (#aggfn)(e, nxt.1);
-                    ht
-                }).into_iter()
-            };
+        let groupbydata_ident = wc.make_ident("groupbydata");
+
+        let (write_prologue, write_iterator) = match persistence {
```

Comment:
could there be/are there any plans for other kinds of persistence in the future other than 'tick and 'static?

Reply:
No plans, tick and static should be enough but we might have more. we might also have less (none) depending on how Mae's idea works out